### PR TITLE
Fix compatibility with libargon2 20171227 and upgrade to Dune

### PR DESCRIPTION
**Pull request by dra27**
- libargon2 20171227 switched the library to `.so.n` format but also bumped the ABI because Argon2id changed the signature of a `argon2_encodedlen`. Switch to loading `libargon2.so.1`, therefore - no attempt at compatibility, given that the release is 3 years old.
- I've not added support for Argon2id, so the signature of `Argon2.encoded_len` (and `Argon2.kind`) haven't changed
- `examples/examples.ml` was broken by various previous API changes - it now builds again
- Removed unused dependency on ppx_deriving and updated the opam file to 2.0
- Upgraded the build system to Dune (`Makefile` targets basically unchanged `make examples` now actually runs the example program instead)
- Documentation therefore builds with odoc, and links (and the `upload_doc`) target should still work

Installed layout should be identical to the previous topkg version.

There might be a follow-up now that I can switch back to Windows...

### CODE REVIEW

**Comment by dra27** [comment]
This would be a good excuse to release and fix #2 at the same time!

File: argon2.opam [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+depends: [
+  "ocaml" {>= "4.02.3"}
+  "dune"
+  "ctypes" {>= "0.4.1"}
+  "ctypes-foreign"
+  "result"
+]
+build: [
+  ["dune" "subst"] {pinned}
+  ["dune" "build" "-p" name "-j" jobs "@install" "@doc" {with-doc}]
```
Comment (Khady) [inline argon2.opam:17]: is that equivalent to?
```
["dune" "build" "-p" name "-j" jobs]
["dune" "build" "-p" name "-j" jobs "@doc"] {with-doc}
```

**Comment by Khady** [comment]
thanks

File: argon2.opam (hunk 1 above)
Comment (dra27) [reply argon2.opam:17]: The end-result is equivalent, but there's only ever one invocation of dune, so it's faster (academically faster for a small library, of course). If a predicate isn't satisfied for a parameter, then it's omitted.

Comment (dra27) [reply argon2.opam:17]: However, you remind me that I've made a mistake - `"odoc" {with-doc}` should be in the `depends` now

**Pull request merged**